### ISSUE TITLE
Slow framerate on Windows 10

### ISSUE THREAD

**Issue by Linkaan** [gonutz/ld46#1]
Video showcasing the problem: https://streamable.com/naz0b7

[keep_the_rectangle_alive.log](https://github.com/gonutz/ld46/files/4535403/keep_the_rectangle_alive.log)

**Comment by Linkaan** [comment]
@gonutz If you would like to get the exact specs of my computer I could also send you a summary of this information.

**Comment by gonutz** [comment]
Does the game consistently start being slow in level 3? Maybe right when the rectangle is in the center of that level? That is where the music starts its infinite loop. The bug could be in the sound, I was looking in the graphics...

This is hard to debug, I added another debug build, more logging. Could you run that as well, same procedure as before? Thanks!

https://github.com/gonutz/ld46/raw/framerate_debugging/keep_the_rectangle_alive.exe

**Comment by Linkaan** [comment]
I tried with the latest version you linked and the game consistently starts being slow at the exact same spot in level 2 it seems (this is also the case in the original build for the jam). I recorded the game from the beginning this time:
https://streamable.com/2xg59h

When I have more time I could try compile the game myself and see if I can profile it myself.

[keep_the_rectangle_alive2.log](https://github.com/gonutz/ld46/files/4540083/keep_the_rectangle_alive2.log)

**Comment by Linkaan** [comment]
I managed to install go on my machine and built the game myself, the issue is still there.

I don't have any experience with go but I installed pprof and modified the main.go file to add cpu profiling. Not sure where to look, but it seems that this line eats up a lot of CPU cycles: https://github.com/gonutz/ld46/blob/c1de7ed97047de47c2ce00f7d3937bd19cdba081/vendor/github.com/gonutz/prototype/draw/window_windows.go#L287

![bild](https://user-images.githubusercontent.com/11377980/80411601-bd5e4800-88cc-11ea-8074-d19da6e1b4cb.png)

I attached files I changed in the zip.
[profile.zip](https://github.com/gonutz/ld46/files/4541635/profile.zip)

**Comment by gonutz** [comment]
Wow, how cool is that, thanks very much! This matches the last log that you sent me. It has a special timer around that exact statement because I was afraid it would be the problem.

The thing is, everything here is hand-written and based on DirectX. Graphics are done with Direct3D9 to be most backwards compatible. The [Present](https://docs.microsoft.com/en-us/windows/win32/api/d3d9/nf-d3d9-idirect3ddevice9-present) method is a call to DirectX that goes into your graphics driver's DLL. There is no way to debug it or know for sure what it does except for disassembling the DLL and stepping through it in a debugger. For that you would need the exact graphics card and driver installed I guess. Can you tell me your graphics card model and version and the driver that you are using?

**Comment by Linkaan** [comment]
https://www.techpowerup.com/gpu-specs/gigabyte-rx-vega-56-gaming-oc.b5106

![bild](https://user-images.githubusercontent.com/11377980/80428681-b9412300-88ea-11ea-95a0-c164fa2f7b11.png)


Happy to help, if there is something else that you want me to try I can do that. I still think it's strange that your old games work fine if they are based on the same engine.

**Comment by Linkaan** [comment]
@gonutz Just a quick update. I tried downgrading my drivers to much older drivers (crimson 17.11.1) and this has completely solved the issue. No problems anymore:
![bild](https://user-images.githubusercontent.com/11377980/80762785-817de980-8b3d-11ea-90a6-32791412275e.png)

I also updated my rating to your game now that I was actually able to play all the levels instead of the first 2 🤣 I really enjoyed the game mechanics that were introduced gradually. The difficulty curve was not linear though, the levels at the end were much easier. The last level reminds me a lot of the flash game Platform Racing 2 where you could create autoplay levels (no player input).

**Comment by gonutz** [comment]
What?! DOWNgrading the driver, huh? That is really strange. I am using the old fixed-function pipeline in the library, maybe this is not really well tested on the latest drivers anymore... The problem with debugging this is that I do not own this graphics card. I would have to have that and the driver to really test this. Google does not help much, the latest information about Direct3D9 is from around 2005 usually :-)

I thought using the oldest, most common DirectX was fine and it seems to be for most users (you are the first one with such a problem in all my LDs) but maybe I should try and have some DirectX10/11/12/99 in there and choose which one to use at runtime depending on the Windows version. I want to stay backward compatible but also there should be no such problems.

Maybe at some point in the future I will create another DirectX wrapper for a newer version. In that case I might come back to this issue and give your future you another version of this game :-)

Thanks for your feedback on the game itself. I actually feel that I have not yet fully explored all the possibilities arising from these simple machanics. I might continue working on this a little more. For example the level where it is all blocks and you just move one out of the way, I liked that and have other ideas with it but I had no time during the event to fully explore that.
I have not heard of Platform Racing 2, it has the aesthetics of a jam game as well :-)

**Comment by Linkaan** [comment]
@gonutz so since I have been having some other strange issues such as this one:
![bild](https://user-images.githubusercontent.com/11377980/80800471-d9523a00-8ba9-11ea-9129-f6c6db170dc8.png)
I decided to run [Guru3D.com]-DDU tool to completely remove the graphics drivers and then reinstall the Adrenalin 20.2.2 driver and guess what, **the issue is no more!**

Ironically when I launched your game the first time after the driver update the game was running at super speed! Did I press some debug button to increase time scale? I think it might have been because I started my computer without having any headphones or speakers connected? Pressing ALT+TAB when in fullscreen maybe? I started the game again after plugging in my headphones and the game played smooth.

Either my GPU is defect somehow or the graphics drivers were corrupt. There was no obvious sign that something was wrong except that I had some issues playing VR games and they crashed at random intervals. 

What I still can't explain is why level 1&2 in your game works completely smooth and then I get these framerate problems. I'm glad it's probably not something that will affect others, just a weird issue on my system.

**Comment by gonutz** [comment]
This is a really interesting endavor you are undertaking :-)

My game uses VSync and relies on it being enabled. Maybe your graphics driver did not really adhere to it? That is actually on my TODO list, cleaning up the code that manages the frame rate. There were issues in the past with 120 Hz monitors and they were fixed. But then maybe there are still more bugs and there is still a TODO in there where I explicitely mention that graphics cards might pretend to vsync on 60 Hz but really they do not. Might be this issue.

Just to get it right, you now eventually re-installed your original driver that did not work. Then it made the game too fast once. But now it works?

**Comment by Linkaan** [comment]
Yes, I completely removed all my old installed graphics drivers using DDU and then I re-installed the original driver that I was having issues with. Now it works fine. I could probably reproduce the increased time scale issue and record it if necessary :p

**Comment by gonutz** [comment]
What is wrong with computers these days? That is why we can't have nice things. But thanks for the detailed report and the work you put in. I really appreciate it, very cool of you! If I can be of any help with one of your projects, feel free to ask :-)

As for the speedy issue, as I said, I have a good idea of the problem and also of the solution, I just have not gotten around to fixing it. But really all that is necessary is a timer that looks to see if 60 Hz between two frames has really passed and wait if not. So you do not need to reproduce this right now. Anyway, once I have that implemented I might come back to you because you have your computer set up the right (wrong) way to test this issue ;-)

**Comment by Linkaan** [comment]
Thanks for keeping the art of creating your own game engine alive, I really admire that! 👍

**Comment by gonutz** [comment]
Thanks! :-)
